### PR TITLE
Logout Refactor

### PR DESCRIPTION
**Pull request by Tishasoumya-02**
#4460

### CODE REVIEW

File: cypress/tests/core/basic/logout.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    const password = ploneAuth[1];
+
+    cy.get('#login').type(user).should('have.value', user);
+    cy.get('#password').type(password).should('have.value', password);
+    cy.get('#login-form-submit').click();
+    cy.get('body').should('have.class', 'has-toolbar');
+  });
+  it('As registered user I can logout', function () {
+    cy.get('#toolbar-personal').click();
+    cy.get('#toolbar-logout').click();
```
Comment (nileshgulia1) [inline cypress/tests/core/basic/logout.js:17]: Maybe just assert if we have `auth_token`? 
https://docs.cypress.io/api/commands/getcookie

File: src/components/theme/Logout/Logout.jsx [hunk 2]
```diff
@@ -26,83 +20,53 @@ const messages = defineMessages({
...
-  /**
-   * Component will receive props
-   * @method componentWillReceiveProps
-   * @param {Object} nextProps Next properties
-   * @returns {undefined}
-   */
-  UNSAFE_componentWillReceiveProps(nextProps) {
-    if (!nextProps.token) {
-      this.props.history.replace(this.props.returnUrl || '/');
+    if (token) {
```
Comment (nileshgulia1) [inline src/components/theme/Logout/Logout.jsx:43]: You can split this into a different useEffect to keep it straightforward and separate concerns.

File: src/components/theme/Logout/Logout.jsx [hunk 3]
```diff
@@ -26,83 +20,53 @@ const messages = defineMessages({
...
     }
-  }
+  }, [dispatch, history, returnUrl, intl, token]);
+
+  return <Login location={{ query: location.query }} />;
+};
+
+Logout.propTypes = {
+  logout: PropTypes.func.isRequired,
+  purgeMessages: PropTypes.func.isRequired,
```
Comment (nileshgulia1) [inline src/components/theme/Logout/Logout.jsx:63]: Unnecessary props?

File: src/components/theme/Logout/Logout.test.jsx [hunk 4]
```diff
@@ -24,7 +25,11 @@ describe('Logout', () => {
     const component = renderer.create(
       <Provider store={store}>
         <IntlProvider locale="en">
-          <Logout location={{ pathname: '' }} />
+          <Logout
+            location={{ pathname: '' }}
+            purgeMessages={purgeMessages}
```
Comment (nileshgulia1) [inline src/components/theme/Logout/Logout.test.jsx:30]: https://github.com/plone/volto/pull/4860/files#r1233651508

**Review by JeffersonBledsoe — CHANGES_REQUESTED** [review]
@Tishasoumya-02 Could we get a test/ include as part of one of the tests you have written already a check for the `returnUrl` feature of this component?

**Comment by Tishasoumya-02** [comment]
Sorry @JeffersonBledsoe  , I am not able to understand

**Comment by JeffersonBledsoe** [comment]
> Sorry @JeffersonBledsoe , I am not able to understand

@Tishasoumya-02 Your current code looks like this:

```jsx
useEffect(() => {
    dispatch(logout());
    dispatch(purgeMessages());
});
```

This code means that the effect will run every time the component renders (more info in the [React docs for useEffect](https://react.dev/reference/react/useEffect#useeffect)). You probably want something like this which has the array of dependencies for the effect: 

```jsx
useEffect(() => {
    dispatch(logout());
    dispatch(purgeMessages());
}, [dispatch]);
```

While it's unlikely that this component will be around long enough for the effect to run multiple times, it is a possibility(and is even something [React will do in dev mode](https://react.dev/reference/react/StrictMode#fixing-bugs-found-by-double-rendering-in-development)).

File: src/components/theme/Logout/Logout.jsx [hunk 5]
```diff
@@ -1,19 +1,12 @@
...
+import { useHistory } from 'react-router-dom';
+import { defineMessages, useIntl } from 'react-intl';
 import qs from 'query-string';
-
-import { Login } from '@plone/volto/components';
+import { Login, Toast } from '@plone/volto/components';
 import { logout, purgeMessages } from '@plone/volto/actions';
 import { toast } from 'react-toastify';
-import { Toast } from '@plone/volto/components';
+import { useToken } from '@plone/volto/hooks/userSession/useToken';
```
Comment (nileshgulia1) [inline src/components/theme/Logout/Logout.jsx:9]:
```suggestion
import { useToken } from '@plone/volto/hooks';
```

**Review by nileshgulia1 — APPROVED** [review]

File: src/components/theme/Logout/Logout.jsx [hunk 6]
```diff
@@ -26,83 +19,44 @@ const messages = defineMessages({
...
+    [location],
+  );
+
+  useEffect(() => {
+    dispatch(logout());
+    dispatch(purgeMessages());
+  }, [dispatch]);
+
+  useEffect(() => {
+    if (token) {
```
Comment (nileshgulia1) [inline src/components/theme/Logout/Logout.jsx:44]: !token?

**Review by nileshgulia1 — CHANGES_REQUESTED** [review]

File: src/hooks/userSession/useToken.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+/**
+ * useToken hook
+ *
+ * This hook returns the current token that is stored in the Redux store in the
+ * `userSession` reducer, and returns it along with the related state (token).
+ *
+ * @export
+ * @return {{ token }}
+ */
+export function useToken() {
```
[the remark this answers is not in the document]
Comment (nileshgulia1) [reply src/hooks/userSession/useToken.js:12]: I see you are accessing directly from redux state in your component. We can remove this!

**Review by nileshgulia1 — CHANGES_REQUESTED** [review]

File: src/hooks/userSession/useToken.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tishasoumya-02) [reply src/hooks/userSession/useToken.js:12]: Yes sure, I might have missed it

**Review by sneridagh — APPROVED** [review]

**Pull request merged**